### PR TITLE
Fix properties of round robin repartition exchange

### PR DESCRIPTION
**Pull request by benben**
Hey partypeople,

We are running a fork of trino at Posthog and merging in upstream changes daily. We saw some of our queries fail after commit c025bf4e8634de26ae5814623e7dfc4ed121e08d (#31128). This is the fix my agent came up with. It is already running on our production env and validated working.

Please let me know if I should change anything!

Thanks! ❤️ 

This is the PR description the agent came up with:

## Description

A round robin repartition exchange (`FIXED_ARBITRARY_DISTRIBUTION`) carries no
partitioning arguments. `PropertyDerivations.visitExchange` derived
`partitionedOn(<that partitioning>)` for it, and `Partitioning.isPartitionedOn(columns)`
iterates the arguments, so with zero arguments it returns `true` for any set of
columns — the exchange claimed to be partitioned on every column.

An aggregation above such an exchange then treated its input as already
partitioned on the grouping keys and skipped the required remote exchange, so
the final aggregation ran independently on every task.

Derive `arbitraryPartition()` instead, for `FIXED_ARBITRARY_DISTRIBUTION` and
`SCALED_WRITER_ROUND_ROBIN_DISTRIBUTION`.

## Additional context and related issues

The bug became reachable with c025bf4e8634de26ae5814623e7dfc4ed121e08d
(#31128), which spreads a single-node cross join probe with a round robin
exchange. That is the first place where these derived properties propagate
upwards: `visitSort` wraps its round robin exchange in a merging exchange, and
the union path in `AddExchanges` builds `PlanWithProperties` without deriving
properties.

Reproducer, which returns two rows per task instead of two rows in total:

    SELECT s.k, COUNT(*) AS n
      FROM UNNEST(SEQUENCE(1, 1000)) AS a(x)
      CROSS JOIN UNNEST(SEQUENCE(1, 1000)) AS b(y)
      CROSS JOIN (
          SELECT z % 2 AS k
          FROM UNNEST(SEQUENCE(1, 10)) AS c(z)
          GROUP BY z % 2
      ) AS s
      GROUP BY s.k
      ORDER BY s.k

The plan shows the missing exchange:

    Fragment 2 [ROUND_ROBIN]
        Aggregate[type = FINAL, keys = [gid]]        <- no remote repartition on gid
        └─ LocalExchange[partitioning = HASH, arguments = [gid::bigint]]
           └─ Aggregate[type = PARTIAL, keys = [gid]]
              └─ CrossJoin[] Distribution: REPLICATED

The probe in the engine test needs more than one page of rows, otherwise the
round robin exchange sends everything to a single task and the bug stays
hidden. `TestAddExchangesPlans.testRemoteExchangeAboveRedistributedCrossJoinProbe`
pins the exchange in the plan itself, so the coverage does not depend on UNNEST
lacking a row estimate.

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
## Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: testing/trino-tests/src/test/java/io/trino/tests/TestDistributedEngineOnlyQueries.java [hunk 1]
```diff
@@ -43,4 +45,25 @@ protected QueryRunner createQueryRunner()
...
+                CROSS JOIN UNNEST(SEQUENCE(1, 10)) AS b(y)
+                CROSS JOIN (
+                    SELECT z % 2 AS k
+                    FROM UNNEST(SEQUENCE(1, 10)) AS c(z)
+                    GROUP BY z % 2
+                ) AS s
+                GROUP BY s.k
+                ORDER BY s.k
+                """))
+                .matches("VALUES (BIGINT '0', BIGINT '10000'), (BIGINT '1', BIGINT '10000')");
```
Comment (raunaqmorarka) [inline testing/trino-tests/src/test/java/io/trino/tests/TestDistributedEngineOnlyQueries.java:66]: The redistribution only fires here because UNNEST has no row-count estimate. `shouldDistributeCrossJoin` returns true on a NaN estimate, and this query's 20000 output rows sit far below the 10000000 default of `min_input_rows_per_task`. The day UNNEST gets a row estimate, the round robin exchange leaves this plan and the test passes with or without the fix.

A plan test next to `testRedistributeSingleNodeCrossJoinProbe` pins the exchange itself and runs in trino-main:

```java
@Test
public void testRemoteExchangeAboveRedistributedCrossJoinProbe()
{
    // the redistributed probe makes the join output arbitrarily partitioned, so a grouped aggregation above it needs a remote exchange
    assertDistributedPlan(
            "SELECT r.regionkey, count(*) FROM (SELECT * FROM nation LIMIT 5) n, region r WHERE n.nationkey < r.regionkey GROUP BY r.regionkey",
            smallCrossJoinRedistribution(),
            anyTree(
                    exchange(REMOTE, REPARTITION, FIXED_HASH_DISTRIBUTION,
                            anyTree(
                                    join(INNER, builder -> builder
                                            .distributionType(REPLICATED)
                                            .left(
                                                    exchange(REMOTE, REPARTITION, FIXED_ARBITRARY_DISTRIBUTION,
                                                            anyTree(
                                                                    tableScan("nation"))))
                                            .right(
                                                    anyTree(
                                                            exchange(REMOTE,
                                                                    REPLICATE,
                                                                    tableScan("region")))))))));
}
```

That one passes with this change and fails with the round robin branch removed from `deriveRepartitionGlobalProperties`.

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PropertyDerivations.java [hunk 2]
```diff
@@ -742,6 +745,18 @@ else if (inputProperties.stream().anyMatch(ActualProperties::isSingleNode)) {
             };
         }
 
+        private static Global deriveRepartitionGlobalProperties(PartitioningScheme partitioningScheme)
+        {
+            Partitioning partitioning = partitioningScheme.getPartitioning();
+            // Round robin partitioning has no partitioning arguments, which would otherwise be interpreted as
+            // being partitioned on every set of columns and let a downstream operator skip a required exchange
+            if (partitioning.getHandle().equals(FIXED_ARBITRARY_DISTRIBUTION) || partitioning.getHandle().equals(SCALED_WRITER_ROUND_ROBIN_DISTRIBUTION)) {
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PropertyDerivations.java]: `StreamPropertyDerivations.visitExchange` has the same special case one level down, but only for `FIXED_ARBITRARY_DISTRIBUTION`, so a local scaled writer exchange (`AddLocalExchanges` line 722) still reports `Optional.of(ImmutableList.of())` partitioning columns and `StreamProperties.isPartitionedOn` is then true for every column set. Nothing above a table writer asks for it today. Adding the handle there too would keep the two derivations in step.

File: testing/trino-tests/src/test/java/io/trino/tests/TestDistributedEngineOnlyQueries.java [hunk 3]
```diff
@@ -43,4 +45,25 @@ protected QueryRunner createQueryRunner()
...
+                SELECT s.k, COUNT(*) AS n
+                FROM UNNEST(SEQUENCE(1, 1000)) AS a(x)
+                CROSS JOIN UNNEST(SEQUENCE(1, 10)) AS b(y)
+                CROSS JOIN (
+                    SELECT z % 2 AS k
+                    FROM UNNEST(SEQUENCE(1, 10)) AS c(z)
+                    GROUP BY z % 2
+                ) AS s
+                GROUP BY s.k
+                ORDER BY s.k
```
Comment (raunaqmorarka) [inline testing/trino-tests/src/test/java/io/trino/tests/TestDistributedEngineOnlyQueries.java]: `matches` compares unordered unless `ordered()` is added, so `ORDER BY s.k` asserts nothing and can go.

**Review by raunaqmorarka — COMMENTED** [review]
The regression this fixes is not in any release (483 predates #31128), so a release note entry would describe a bug no user can hit. Dropping the proposed note keeps the 484 notes to the change users see.

**Comment by raunaqmorarka** [comment]
@benben could you please send in your CLA ?
cc: @martint

**Comment by benben** [comment]
@raunaqmorarka thanks for you feedback! I pushed all the fixes as requested. I sent the cla PDF yesterday right after my PR from the same email I pushed these commits.

**Review by raunaqmorarka — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PropertyDerivations.java [hunk 4]
```diff
@@ -742,6 +751,18 @@ else if (inputProperties.stream().anyMatch(ActualProperties::isSingleNode)) {
             };
         }
 
+        private static Global deriveRepartitionGlobalProperties(PartitioningScheme partitioningScheme)
+        {
+            Partitioning partitioning = partitioningScheme.getPartitioning();
+            // Round robin partitioning has no partitioning arguments, which would otherwise be interpreted as
+            // being partitioned on every set of columns and let a downstream operator skip a required exchange
+            if (isRoundRobin(partitioning.getHandle())) {
+                return arbitraryPartition();
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PropertyDerivations.java]: The guard enumerates handles, but the property being guarded against is the shape. `Partitioning.isPartitionedOn` and `isEffectivelySinglePartition` return true for any partitioning with zero arguments, and `isCompatibleWith` treats two zero-argument partitionings on the same handle as compatible. Any future handle without arguments, or a connector `ConnectorTablePartitioning` that comes back with an empty column list, would take the same path through `visitTableScan` (line 932 area) and `planPartitionedJoin`. Consider putting the check in `Partitioning.isPartitionedOn` instead, returning `handle.isSingleNode()` when `arguments` is empty, so SINGLE and COORDINATOR_ONLY keep their "partitioned on everything" meaning and every other argument-free partitioning reads as arbitrary. That would also let `isRoundRobin` go away.

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StreamPropertyDerivations.java [hunk 5]
```diff
@@ -376,7 +376,9 @@ public StreamProperties visitExchange(ExchangeNode node, List<StreamProperties>
 
             return switch (node.getType()) {
                 case GATHER -> StreamProperties.singleStream();
-                case REPARTITION -> node.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_ARBITRARY_DISTRIBUTION) ?
+                // Round robin partitioning has no partitioning columns, which would otherwise be interpreted
+                // as being partitioned on every set of columns
+                case REPARTITION -> isRoundRobin(node.getPartitioningScheme().getPartitioning().getHandle()) ?
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StreamPropertyDerivations.java]: Same point here. `getPartitioning().getArguments().isEmpty()` describes the condition the comment states, and covers a handle this list does not name. Remote exchanges never reach this branch, so the check only sees local round robin today, which makes the shape test safe to use.

**Review by raunaqmorarka — COMMENTED** [review]
The fix is correct for the two handles it names. Two suggestions below on making the guard structural, so the next argument-free partitioning does not regress the same way.

**Comment by benben** [comment]
@raunaqmorarka addressed your feedback!

**Review by raunaqmorarka — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StreamPropertyDerivations.java [hunk 6]
```diff
@@ -376,7 +375,9 @@ public StreamProperties visitExchange(ExchangeNode node, List<StreamProperties>
 
             return switch (node.getType()) {
                 case GATHER -> StreamProperties.singleStream();
-                case REPARTITION -> node.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_ARBITRARY_DISTRIBUTION) ?
+                // A partitioning without columns would otherwise be interpreted as being partitioned
+                // on every set of columns
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StreamPropertyDerivations.java:379]: Partitioning on zero columns means all rows belong to the same partition. I.e., it has the same co-location guarantee as a gather. An empty routing argument list doesn't imply this guarantee. Round-robin also has no arguments.

File: core/trino-main/src/main/java/io/trino/sql/planner/Partitioning.java [hunk 7]
```diff
@@ -159,6 +159,12 @@ private static boolean isPartitionedWith(
 
     public boolean isPartitionedOn(Collection<Symbol> columns, Set<Symbol> knownConstants)
     {
+        if (arguments.isEmpty()) {
+            // A partitioning without arguments is partitioned on every set of columns only when all rows end up in
+            // a single partition, as they do for SINGLE and COORDINATOR_ONLY. Round robin and every other
+            // argument-free partitioning spread rows across partitions independently of any column.
+            return handle.isSingleNode();
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/Partitioning.java:166]: This loses valid partitioning guarantees:

- Zero-key `FIXED_HASH_DISTRIBUTION` sends all rows to one partition (`InterpretedHashGenerator:102`). Its handle still returns false from `isSingleNode()`.
- An empty column list means the table is entirely on one worker (`ConnectorTablePartitioning:75`). The default `ConnectorPartitioningHandle.isSingleNode()` returns false (`ConnectorPartitioningHandle:24`).

Same issue in `StreamPropertyDerivations:380`.

Also, `isPartitionedOnExactly` still returns true for round-robin with an empty key set (`Partitioning:184`). With `use_exact_partitioning=true`, `AddExchanges:388` can skip the gather for a window without `PARTITION BY`. The checks for pattern recognition (`AddExchanges:419`) and top-N ranking (`AddExchanges:541`) have the same problem.

Remove the empty-argument guards. Derive `arbitraryPartition()` for round-robin in `PropertyDerivations:732`. Use `Optional.empty()` for its local stream properties in `StreamPropertyDerivations:380`. Keep the single-partition meaning of a known empty key set. That fixes both property checks without discarding the guarantees above.

Add tests for zero-key hash partitioning and connector partitioning with empty columns and the default `isSingleNode()`. Also, test a window without `PARTITION BY` above the redistributed cross-join probe with `use_exact_partitioning=true`.

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/Partitioning.java [hunk 8]
```diff
@@ -159,6 +159,12 @@ private static boolean isPartitionedWith(
...
+        }
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/Partitioning.java]: Remove this guard and derive the round robin case in `PropertyDerivations.visitExchange` instead, so zero-key hash and connector partitionings keep their single-partition meaning.

```suggestion
```

The `PropertyDerivations` side, which also needs imports for `Partitioning`, `PartitioningHandle`, `PartitioningScheme` and the two round robin handles:

```java
    // PropertyDerivations
    static boolean isRoundRobin(PartitioningHandle handle)
    {
        return handle.equals(FIXED_ARBITRARY_DISTRIBUTION) || handle.equals(SCALED_WRITER_ROUND_ROBIN_DISTRIBUTION);
    }

    // PropertyDerivations.Visitor.visitExchange
                case REPARTITION -> ActualProperties.builder()
                        .global(deriveRepartitionGlobalProperties(node.getPartitioningScheme()))
                        .constants(constants)
                        .build();

    // PropertyDerivations.Visitor
        private static Global deriveRepartitionGlobalProperties(PartitioningScheme partitioningScheme)
        {
            Partitioning partitioning = partitioningScheme.getPartitioning();
            // round robin places rows independently of their values
            if (isRoundRobin(partitioning.getHandle())) {
                return arbitraryPartition();
            }
            return partitionedOn(partitioning)
                    .withReplicatedNulls(partitioningScheme.isReplicateNullsAndAny());
        }
```

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StreamPropertyDerivations.java [hunk 9]
```diff
@@ -376,7 +375,9 @@ public StreamProperties visitExchange(ExchangeNode node, List<StreamProperties>
...
+                case REPARTITION -> node.getPartitioningScheme().getPartitioning().getArguments().isEmpty() ?
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/StreamPropertyDerivations.java]: Check only the round robin handles here, with a static import of `PropertyDerivations.isRoundRobin` in place of the removed `FIXED_ARBITRARY_DISTRIBUTION` import.

```suggestion
                case REPARTITION -> isRoundRobin(node.getPartitioningScheme().getPartitioning().getHandle()) ?
```

File: core/trino-main/src/main/java/io/trino/sql/planner/Partitioning.java [hunk 10]
```diff
@@ -159,6 +159,12 @@ private static boolean isPartitionedWith(
 
     public boolean isPartitionedOn(Collection<Symbol> columns, Set<Symbol> knownConstants)
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/Partitioning.java]: Add a `TestPartitioning` next to this class, whose two tests fail with the guard below and pass without it.

```java
package io.trino.sql.planner;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import io.trino.spi.connector.ConnectorPartitioningHandle;
import io.trino.spi.connector.ConnectorTransactionHandle;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static io.trino.spi.type.BigintType.BIGINT;
import static io.trino.sql.planner.SystemPartitioningHandle.FIXED_HASH_DISTRIBUTION;
import static io.trino.testing.TestingHandles.TEST_CATALOG_HANDLE;
import static org.assertj.core.api.Assertions.assertThat;

public class TestPartitioning
{
    private static final Symbol COLUMN = new Symbol(BIGINT, "column");

    @Test
    public void testHashPartitioningWithoutColumns()
    {
        // hashing zero columns sends every row to the same partition
        Partitioning partitioning = Partitioning.create(FIXED_HASH_DISTRIBUTION, ImmutableList.of());
        assertThat(partitioning.isPartitionedOn(ImmutableList.of(COLUMN), ImmutableSet.of())).isTrue();
    }

    @Test
    public void testConnectorPartitioningWithoutColumns()
    {
        // a connector partitioning without columns keeps the whole table on one worker
        PartitioningHandle handle = new PartitioningHandle(
                Optional.of(TEST_CATALOG_HANDLE),
                Optional.of(new ConnectorTransactionHandle() {}),
                new ConnectorPartitioningHandle() {});
        Partitioning partitioning = Partitioning.create(handle, ImmutableList.of());
        assertThat(partitioning.isPartitionedOn(ImmutableList.of(COLUMN), ImmutableSet.of())).isTrue();
    }
}
```

File: core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestAddExchangesPlans.java [hunk 11]
```diff
@@ -251,6 +251,29 @@ public void testRedistributeSingleNodeCrossJoinProbe()
...
+                                                .left(
+                                                        exchange(REMOTE, REPARTITION, FIXED_ARBITRARY_DISTRIBUTION,
+                                                                anyTree(
+                                                                        tableScan("nation"))))
+                                                .right(
+                                                        anyTree(
+                                                                exchange(REMOTE,
+                                                                        REPLICATE,
+                                                                        tableScan("region")))))))));
+    }
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestAddExchangesPlans.java]: Add the window case with `use_exact_partitioning` enabled, which also needs an import of `WindowNode`. At this head the window runs over the round robin join output with no gather, so this test fails there and passes with the round robin derivation.

```java
    public void testGatherAboveRedistributedCrossJoinProbeWithExactPartitioning()
    {
        // a window without PARTITION BY above the redistributed probe needs a gather
        assertDistributedPlan(
                "SELECT sum(r.regionkey) OVER () FROM (SELECT * FROM nation LIMIT 5) n, region r WHERE n.nationkey < r.regionkey",
                Session.builder(smallCrossJoinRedistribution())
                        .setSystemProperty(USE_EXACT_PARTITIONING, "true")
                        .build(),
                anyTree(
                        node(WindowNode.class,
                                anyTree(
                                        exchange(REMOTE, GATHER,
                                                anyTree(
                                                        join(INNER, builder -> builder
                                                                .distributionType(REPLICATED)
                                                                .left(
                                                                        exchange(REMOTE, REPARTITION, FIXED_ARBITRARY_DISTRIBUTION,
                                                                                anyTree(
                                                                                        tableScan("nation"))))
                                                                .right(
                                                                        anyTree(
                                                                                exchange(REMOTE,
                                                                                        REPLICATE,
                                                                                        tableScan("region")))))))))));
    }
```

File: core/trino-main/src/main/java/io/trino/sql/planner/Partitioning.java (hunk 7 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/Partitioning.java]: Agreed, the empty-argument guard I suggested earlier is the wrong layer. At this head the window without `PARTITION BY` plans no gather above the round robin join output. The inline comments have the round robin derivation and the tests you asked for, which fail at this head and pass with it.

**Comment by benben** [comment]
thanks for the feedback! should have addressed all.

**Comment by raunaqmorarka** [comment]
Thanks, this looks good to me. Could you squash the two commits into one? The second commit removes the `Partitioning` guard that the first one adds. The first commit's subject still describes the combined change.

**Comment by benben** [comment]
> Thanks, this looks good to me. Could you squash the two commits into one? The second commit removes the `Partitioning` guard that the first one adds. The first commit's subject still describes the combined change.

done!

**Review by raunaqmorarka — APPROVED** [review]

**Review by martint — APPROVED** [review]
